contracts/finality/src/state/public_randomness.rs

```diff
@@ -101,3 +102,98 @@ pub fn get_pub_rand_commit(
     // Return the results or an empty vector if no results found
     Ok(res)
 }
+
+/// Inserts a public randomness value into the PUB_RAND_VALUES map for the given fp and height.
+///
+/// - If no value exists, it inserts the value.
+/// - If the same value already exists, it is a no-op and returns Ok(())
+/// - If a different value exists, it returns ContractError::PubRandAlreadyExists.
+///   This is an error as the contract should recognize only a single public randomness value
+///   for a specific height per finality provider.
+pub fn insert_pub_rand_value(
+    storage: &mut dyn Storage,
+    fp_btc_pk_hex: &str,
+    height: u64,
+    pub_rand: &[u8],
+) -> Result<(), ContractError> {
+    if let Some(existing) = PUB_RAND_VALUES.may_load(storage, (fp_btc_pk_hex, height))? {
+        if existing == pub_rand {
+            return Ok(());
```

Comment:
It should never happen that a duplicated pub rand is inserted, so if L119 is triggered we could simply return error

Reply:
My thought was that for the context of the function as the gatekeeper of storage, it only cares about replacing a value and only returns an error in that case. wdyt?